### ISSUE TITLE
[VC]: Sync Kubernetes ResourceQuota to Tenant Namespace

### ISSUE THREAD

**Issue by weiling61** [kubernetes-sigs/multi-tenancy#1036]
ResourceQuota restricts tenant from using too much resources in super cluster. Would there be interest in applying resource quota to tenant namespace?  
- First shall we sync resource quota upward to tenant namespace. This way tenant can monitor his/her resource usage and tenant's resource request can be restricted in VC space. 
- Second, if Resource Quota of virtual cluster has been configured, can Syncer apply the same quota or a subset quota to tenant’s namespace?
- Third, in the future, we can have an extended admission controller plugin to monitor/control aggregated resource usage of all VC tenants across their namespaces.

@christopherhein @Fei-Guo

**Comment by christopherhein** [comment]
Thanks for filing this @weiling61 I think syncing ResourceQuotas could be good but it heavily depends on how it was implemented for each supercluster as cluster admins, similarly to the priorityclasses. I could see this being done as one of the optional syncers.

Some notes we need to consider:
* if the tenant control plane has the quota admission controller turned on will it update the `status.used` on the `ResourceQuota` and how will, assuming we're not going to be syncing downward we'll need to make sure that those stay in sync somehow.
* Because we want to stay true to core kubernetes implementation if we sync them I think we need to make sure we're syncing the entire object and not a subset quota like you called out

**Comment by weiling61** [comment]
If in sync is required, we probably only allow tenants to monitor their quota usages, but not allow them to modify them
I did not explain clearly about "the subset of quota" idea. Yes we will always sync entire quota object to tenant namespace. But we can choose a smaller quota value if needed.

**Comment by christopherhein** [comment]
I see, thanks for the clarification. The smaller quota value seems like it would be out of the scope for syncing resources since it might be used in different ways depending on how someone is using and deploying virtualcluster.

**Comment by weiling61** [comment]
Agree, small value is an example

**Comment by charleszheng44** [comment]
@christopherhein @weiling61  I agree. How to synce resource quota or priority class really depends on how supercluster implements the resource management mechanism for vc. 

In my opinion, maybe we can make the resource quota an optional resource for syncing. Or more flexibly, provide new syncer options like `--uws-resources`, `--dws-resources`, which allow the user to decide how they want to sync the resource.

**Comment by christopherhein** [comment]
> In my opinion, maybe we can make the resource quota an optional resource for syncing. Or more flexibly, provide new syncer options like `--uws-resources`, `--dws-resources`, which allow the user to decide how they want to sync the resource.

Sounds like a good idea, from your experiences, do you see the syncers as being independent enough where you could have that level of options, ie could you implement dws without uws or vice versa and it wouldn't cause issues? or is that a lot more effort to maintain?

**Comment by Fei-Guo** [comment]
In many cases, tenants want to control their quota independently in tenant master. Normally super does not need to be aware of tenant quota for function correctness (I do see request to sync them because some monitoring tools may watch super master only for tenant quota usage though). I tend not to sync quota from tenant to super. 

Syncing from super to tenant is interesting. This is something that tenant is not expected, hence it is less likely tenant has any tool depending on the quota object. In this case, the difference between syncing and not syncing quota is that, the former may prevent Pod objects from creating and the latter may make the Pod stuck in pending status for a long time (syncer will keep retry until some other tenant Pods are gone which free the quota). In the latter case, we can also send event to tenant to notify user that the quota in super is used up.  Overall, the question is that which is the preferred behavior or can the behavior of not syncing quota be acceptable.

I don't see a strong need to sync quota resource. If it has to be synced in some cases I think the source of truth should be tenant.  And one important thing is that we should keep the Pod resource configuration consistent between tenant and super if resource quota is synced. This means any sidecar webhook cannot be installed in super and it has to be installed in each tenant.

**Comment by Fei-Guo** [comment]
> > In my opinion, maybe we can make the resource quota an optional resource for syncing. Or more flexibly, provide new syncer options like `--uws-resources`, `--dws-resources`, which allow the user to decide how they want to sync the resource.
> 
> @charleszheng44 Sounds like a good idea, from your experiences, do you see the syncers as being independent enough where you could have that level of options, ie could you implement dws without uws or vice versa and it wouldn't cause issues? or is that a lot more effort to maintain?

In my opinion the resource sync direction has to be deterministic and predefined (the other direction should not be supported). Note that the syncer work for all tenants, if we open the sync direction flexibility, what if some tenants want one direction but the others want the opposite direction. This is way too complicated to manage.

**Comment by christopherhein** [comment]
> In my opinion the resource sync direction has to be deterministic and predefined (the other direction should not be supported). Note that the syncer work for all tenants, if we open the sync direction flexibility, what if some tenants want one direction but the others want the opposite direction. This is way too complicated to manage.

Yea, this was my concern as well regarding complexity for maintenance.

**Comment by weiling61** [comment]
I agree that resource downward (tenant to super) syncing will bring quite some confusions. Upward syncing (super to tenant) makes more sense. ResourceQuota is an isolation mechanism super wants to use to control tenant behaviors. This is similar to OS system ----  admin can set quotas for user,  user can check the quotas usage but cannot change the quota. 

If super wants “the source of truth should be tenant”, they just do not turn on --extra-syncing-resources=resourcequota and api-server in each vc cluster can handle quota restriction.  If super turns on ResourceQuota syncing. It means tenant should not be allowed to modify assigned quotas at all. Resource quota UWS will over-write the quota modified by tenant, which provides super an enforced resource isolation mechanism.

**Comment by Fei-Guo** [comment]
> I agree that resource downward (tenant to super) syncing will bring quite some confusions. Upward syncing (super to tenant) makes more sense.
> In most cases, ResourceQuota is an isolation mechanism super wants to use to control tenant behaviors. This is similar to OS system ---- admin can set quotas for user, user can check the quotas usage but cannot change the quota.
> 
> If super wants “the source of truth should be tenant”, they just do not turn on --extra-syncing-resources=resourcequota and api-server in each vc cluster can handle quota restriction. If super turns on ResourceQuota syncing. It means tenant should not be allowed to modify assigned quotas at all. Resource quota UWS will over-write the quota modified by tenant, which provides super an enforced resource isolation mechanism.

My question is why we need to sync the quota to tenant? The tenant cannot consume over the super resource quota anyway.  Will sending an event to tell tenant that the Pod cannot be run because it violates the super limitation be enough? More importantly, tenant is fully open to tenant, so any overwrite to tenant created quota resources will be very confusing to the tenant users.

Normally the capacity planning has to be done between cluster-admin and tenant and it is fairly reasonable that tenant choose a smaller assigned quota value in tenant namespace for different reasons. The tenant quota and super quota are not necessary to be identical. 

Another thought is that it is not practical for super to assign quota to each tenant namespace since the number of namespace can grow. It is hard to control overall tenant resource usage by setting the quota per namespace assuming this is the ending goal. Tenant can just bypass the limitation by creating more namespaces. Super needs to implement another webhook to do the check instead of relying on the quota system.

**Comment by weiling61** [comment]
"sync the quota to tenant" from super is just one option. If this use case is not very common, we probably do not need bother to implement resource quota UWS.

Cluster-admin and tenant do their capacity planning/quota management separately without sync is another option.
